File: Controllers/SamplingController.cpp
#include <essentia.h>
#include <json/value.h>
#include <stdbool.h>
#include <vector>
//include json

class SamplingController
{
    public:

    int sampleTime,sampleRate,waitTime,offset;
    int* array;
    bool exit_flag, finished;

    /*
      mainThread
      model
      metadata
    */

    void SamplingController(int sampleTime=15, int sampleRate=48000, string waitTime)
    {
        this.sampleTime = sampleTime;
        this.sampleRate = sampleRate;

        if (waitTime.compare("") != 0)
        {
            this.waitTime = atoi(waitTime);
        }else
        {
            this.waitTime = 45;
        }

        //Include essentia related stuff here

        this.offset = 0;
        this.array = (int*)malloc(this.sampleRate*15*sizeof(int));
    }

    void mainSample()
    {
        this.exit_flag = false;
        while (!this.exit_flag)
        {
            this.finished = false;
            //INCLUDE THREAD CODE HERE
            bool workdone = false;
            while (!this.finished)
            {
                if(!this.offset == 14 && !workdone)
                {
                    this.performAnalysis();
                    workdone = true;
                }
            }
            this.finished = false;
            //INCLUDE THREAD CODE HERE
            while (!this.finished)
            {
                continue;
            }
        }
    }

    void appendAudio(string audioPath)
    {
        if (this.offset == 14)
        {
            this.offset = 0;
        }
        //ESSENTIA LOADER
        //std::vector<int> audio = essentia loader;
        IntegerVector idx = IntegerVector::create(this.offset*this.sampleRate, (this.offset+1)*this.sampleRate);
        //self.array[idx] = audio;
        this.offset++;
    }

    void timer(int timer)
    {
        //c++ threading sleep function here
        this.finished = true;
    }

    int getSampleTime()
    {
        return this.sampleTime;
    }

    int getWaitTime()
    {
        return this.waitTime;
    }

    void setSampleTime(int sampleTime)
    {
        this.sampleTime = sampleTime;
    }

    void setWaitTime(int waitTime)
    {
        this.waitTime = waitTime;
    }

    void performAnalysis()
    {
        //int scaled = this.array;

    }

    std::vector<string> parseTags(std::vector<double> activations)
    {
        std::vector<string> tags;
        int count = 0;
        for (int i = 0; i < this.metadata.length; i++)
        {
            double probability = activations[i];
            string label = this.metadata[i];
            if ((int) ((float) probability * 100) > 50)
            {
                tags.push_back(parseTag(label));
            }
        }
        return tags;
    }

    string parseTag(string label)
    {
        switch(label)
        {
            case "blu":
                return "Blues";
            case "cla":
                return "Classic";
            case "cou":
                return "Country";
            case "dis":
                return "Disco";
            case "hip":
                return "Hip Hop";
            case "jaz":
                return "Jazz";
            case "met":
                return "Metal";
            case "pop":
                return "Pop";
            case "reg":
                return "Reggae";
            case "roc":
                return "Rock";
        }
    }
};

File: Controllers/SamplingController.py
import json
import numpy as np
from os import getcwd

from Socket.Socket import *

class SamplingController:

    # Shouldn't ever change
    # KVP since switches aren't a thing
    CONVERSIONS = {
        "blu": "Blues",
        "cla": "Classic",
        "cou": "Country",
        "dis": "Disco",
        "hip": "Hip Hop",
        "jaz": "Jazz",
        "met": "Metal",
        "pop": "Pop",
        "reg": "Reggae",
        "roc": "Rock"
    }

    def __init__(self, core, sampleTime=15, sampleRate=48000):
        self.core = core
        self.sampleRate = sampleRate
        self.sampleTime = sampleTime

        # Set the wait time from configurations (if applicable),
        # use default otherwise.
        waitTime = self.core.getConfiguration("Sampling", "wait", str)
        if waitTime != "":
            self.waitTime = int(waitTime)
        else:
            self.waitTime = 45
        self.metadata = json.load(open('Files\\genre_tzanetakis-musicnn-msd-1.json', 'r'))

        self.socket = Socket(6000, "connect", host='127.0.0.1')

    # Tell essentia main to anaylze the provided file
    def requestPerformAnalysis(self):
        self.socket.send('read')
        self.socket.recv()
        self.socket.send('song.wav')
        self.socket.recv()
        data = self.socket.recv()
        
        activations = data
        while True:
            activations += data
            if data[len(data)-2:len(data)] == "]]":
                break
            data = self.socket.recv()

        
        # Remove OK
        print(activations)
        activations = activations[0:len(activations)-2]

        activations = activations.replace(']', ' ')
        activations = activations.replace('[', ' ')

        count = 0
        activationsLists = []
        current = []
        for item in activations.split(' '):
            if item == '' or item == '\n' or item == '\r':
                continue
            count += 1
            
            current += [item]
            if count % 10 == 0:
                activationsLists += [current]
                current = []

        activationFloats = np.array(activationsLists).astype(np.float32)

        tags = self.parseTags(activationFloats)
        self.core.sendTags(tags)

    # Alert essentia main that app is shutting down
    def requestClose(self):
        self.socket.send('close')
        self.socket.recv()
        self.socket.recv()
        self.socket.close()

    #Returns the sampleTime
    def getSampleTime(self):
        return self.sampleTime

    #Considerations of scrapping this function because the sampleTime will always be 15.
    def setSampleTime(self,sampleTime):
        self.sampleTime = sampleTime

    #waitTime can be configured in the
    def setWaitTime(self,waitTime):
        self.socket.send('set')
        self.socket.recv()
        self.socket.send(str(waitTime))
        self.waitTime = waitTime

    #Returns the waitTime
    def getWaitTime(self):
        return self.waitTime

    # Parse tags provided by essentia
    def parseTags(self, activations):
        tags = []
        #count = 0
        meanActivations = activations.mean(axis=0)
        m = np.argmax(meanActivations, axis=0)
        tags += [self.CONVERSIONS[self.metadata['classes'][m]]]

        #for label, probability in zip(self.metadata['classes'], activations.mean(axis=0)):
            #if int(float(probability) * 100) > 50:
                # Convert from metadata tag into actual tag
                #tags.append(self.CONVERSIONS[label])
            #count+=1
        return tags
